lib/net/dhcp/record.rb

```diff
@@ -52,19 +52,22 @@ def create
       end
     end
 
-    # Returns an array of record objects which are conflicting with our own
+    # Returns an array of record objects which are conflicting with our own. Comparsion is done
+    # only using selected attributes: MAC, IP, hostname and network.
     def conflicts
-      conflicts = [proxy.record(network, mac), proxy.records_by_ip(network, ip)].flatten.compact.delete_if { |c| c.lease? || c == self || related_macs.include?(c.mac) }
+      conflicts = [proxy.record(network, mac), proxy.records_by_ip(network, ip)].flatten.compact.delete_if do |c|
+        c.lease? || c.eql_for_conflicts?(self) || related_macs.include?(c.mac)
+      end
       @conflicts ||= conflicts.uniq {|c| c.attrs}
     end
 
     # Verifies that a record already exists on the dhcp server
     def valid?
       logger.info "Fetching DHCP reservation #{name} for #{self}"
-      self == proxy.record(network, mac)
+      self.eql_for_conflicts?(proxy.record(network, mac))
```

Comment:
I think you may be breaking the API here. This called at:
https://github.com/theforeman/foreman/blob/4485b8e8f92bafc5eb7d3ae5858985d4a8c9079c/app/models/concerns/orchestration/dhcp.rb#L187-L189
Then that's called by
https://github.com/theforeman/foreman/blob/4485b8e8f92bafc5eb7d3ae5858985d4a8c9079c/app/models/concerns/orchestration/dhcp.rb#L171-L185

By making this change, the DHCP orchestration will no longer detect that the filename needs to be updated.

IMHO the `==` method was still the correct one and removing it here is not the right thing to do. Am I missing something?

Reply:
It works, I don't understand.

Reply:
The point I tried to make is that there's 2 flows:
* Is there a conflicting DHCP record? (what you're implementing here)
* Does the existing DHCP record need an update? (what I'm linking in my above comment)

My point is about the latter flow. To spell it out the call graph:
* `dhcp_update_required?` calls `all_dhcp_records_valid?`
* `all_dhcp_records_valid?` checks whether there's any mismatch. Your code change means it ignores the filename

This means you are breaking the API that `all_dhcp_records_valid?` expects. This may not be an issue because there's also a `operatingsystem.boot_filename(old.host) != operatingsystem.boot_filename(self.host)` check.

What I'm suggesting that the API is much cleaner if you have two methods on the DHCP record:
* `conflicts` checks whether there's a conflicting record
* `==(other)` checks if the record is exactly the same. If it's not, it needs an update.

That would make the whole orchestration much easier to follow.